### ISSUE TITLE
The HPOS backfill and sync processes accidentally remove meta data entries saved with save_meta_data() (instead of save())

### ISSUE THREAD

**Issue by wvega** [woothemes/woocommerce#39559]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

The backfill and sync processes for orders accidentally remove meta data entries, because adding meta data to an order (using `update_meta_data()` for example) followed by a call to `save_meta_data()` is not reflected on other order instances retrieved using `wc_get_order()`.

My best guess is that calling `save_meta_data()` is not clearing a cache. 

When COT tables are used, other calls to `wc_get_order()` in the same request return a second order instance that doesn't include the newly added meta data. As a result, when someone calls `save_meta_data()` or `save()` on the second order instance, [the backfill process](https://github.com/woocommerce/woocommerce/blob/3a2922567e864c8ca281801384438253489d979a/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php#L559) removes the newly added metadata from the postmeta table [because it appears as it was deleted from the order](https://github.com/woocommerce/woocommerce/blob/34a8273268efb26a30c0ec4da80e2c51ab8b68cf/plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php#L729-L739).

At this point, the new meta data entries should still show show up in `wc_orders_meta`. However, when someone else tries to read the order, [the sync process removes the newly added meta from `wc_orders_meta`](https://github.com/woocommerce/woocommerce/blob/3a2922567e864c8ca281801384438253489d979a/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php#L1300) becuase there is no matching metadata in the postmeta table.

### Expected behavior

* Calling `save_meta_data()` prevents the backfill and sync process from removing newly added meta data entries
* Order instances returned by `wc_get_order()` include metadata saved using `save_meta_data()` on a different instance that represents the same order

### Actual behavior

Calling `save_meta_data()` or `save()` on an order instance causes meta data previously saved with `save_meta_data()` on a different instance (during the same request) to be removed from the database.

### Steps to reproduce

1. Update the snippet below replacing 26 with the ID of an order in your test site.
1. Run the snippet with WooCommerce configured to use the COT and configured to use WordPress post tables. The `save-meta-data` log file will show that:
  * meta data entries are correctly saved when the WordPress post tables are used 
  * meta data entries are not available on the order instance and are removed from the database when the COT are used

```php
use Automattic\WooCommerce\Utilities\OrderUtil;

$meta_key_one = '_test_meta_data_'.time().'_one';
$meta_key_two = '_test_meta_data_'.time().'_two';

add_action('admin_init', function() use ($meta_key_one, $meta_key_two) {
	if ($order = wc_get_order(26)) {
		$order->update_meta_data($meta_key_one, 123);
		$order->save_meta_data();
	}
}, 10);

add_action('admin_init', function() use ($meta_key_one, $meta_key_two) {
	if ($order = wc_get_order(26)) {
		$order->update_meta_data($meta_key_two, 456);
		
		// save_meta_data() triggers a backfill process that removes $meta_key_one from the postmeta table because it doesn't show up in the order
		$order->save_meta_data();
	}
}, 20);

add_action('admin_init', function() use ($meta_key_one, $meta_key_two) {
	if ($order = wc_get_order(26)) {
		// save() triggers a backfill process that removes $meta_key_two from the postmeta table because it doesn't show up in the order
		$order->save();
		
		$meta_keys = array_map(static fn ($metadata) => $metadata->key, $order->get_meta_data());
		
		wc_get_logger()->debug($cot_enabled.json_encode($meta_keys), ['source' => 'save-meta-data']);		

		$cot_enabled = OrderUtil::custom_orders_table_usage_is_enabled() ? '[COT=on]' : '[COT=off]';

		wc_get_logger()->debug($cot_enabled.(in_array($meta_key_one, $meta_keys, true) ? "{$meta_key_one} is set" : "{$meta_key_one} is NOT set"), ['source' => 'save-meta-data']);
		wc_get_logger()->debug($cot_enabled.(in_array($meta_key_two, $meta_keys, true) ? "{$meta_key_two} is set" : "{$meta_key_two} is NOT set"), ['source' => 'save-meta-data']);
	}
}, 30);
```

### WordPress Environment

<details>
<summary>WooCommerce System Status Report</summary>

### WordPress Environment ###

WordPress address (URL): https://deploy.skyverge.test
Site address (URL): https://deploy.skyverge.test
WC Version: 8.0.0
REST API Version: ✔ 8.0.0
WC Blocks Version: ✔ 10.6.2
Action Scheduler Version: ✔ 3.6.1
Log Directory Writable: ✔
WP Version: 6.2.2
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.23.3
PHP Version: 8.2.1
PHP Post Max Size: 512 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.87.0
(SecureTransport) OpenSSL/1.1.1s

SUHOSIN Installed: –
MySQL Version: 5.7.40
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.0.0
WC Database Prefix: wp_
Total Database Size: 5.99MB
Database Data Size: 4.19MB
Database Index Size: 1.80MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.13MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 3.05MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.03MB + Index: 0.06MB + Engine InnoDB
wp_snippets: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_pickup_locations_geodata: Data: 0.02MB + Index: 0.05MB + Engine InnoDB

### Post Type Counts ###

attachment: 7
page: 7
post: 2
product: 1
shop_order: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (3) ###

Query Monitor: by John Blackbourn – 3.13.1
Code Snippets: by Code Snippets Pro – 3.4.2
WooCommerce: by Automattic – 8.0.0-rc.1

### Dropin Plugins (1) ###

db.php: Query Monitor Database Class (Drop-in)

### Settings ###

API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: –
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: ✔

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twelve
Version: 3.3 (update to version 3.9 is available)
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
product-variation-management
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2023-08-02 22:20:03 +00:00
Options: ✔
Notes: 100
Onboarding: skipped

### Action Scheduler ###

Complete: 40
Oldest: 2023-08-02 21:01:57 +0000
Newest: 2023-08-02 21:53:12 +0000

Pending: 3
Oldest: 2023-08-02 22:34:03 +0000
Newest: 2023-08-03 21:20:42 +0000

### Status report information ###

Generated at: 2023-08-02 22:12:42 +00:00
</details>

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by wvega** [comment]
As mentioned in https://github.com/woocommerce/woocommerce/issues/34981, replacing `save_meta_data()` with `save()` prevents this problem from happening. However, I decided to still open the bug because:

* https://github.com/woocommerce/woocommerce/issues/35703 and other conversations I found state that `save_meta_data()` should be enough to persist meta data changes.
* I couldn't find a recommendation to stop using `save_meta_data()`, even though the [the upgrade recipe book](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#apis-for-gettingsetting-posts-and-postmeta) uses `save()` in the meta data example.

**Comment by vedanshujain** [comment]
This should be addressed by #39450

**Comment by wvega** [comment]
Thank you @vedanshujain!

I just tried again running my snippet with the changes from https://github.com/woocommerce/woocommerce/pull/39450: the new meta data entries are no longer permanently removed, but they are still unavailable in some cases.

I added the following line to my snippet to log the meta keys known to the order instance:

```php
wc_get_logger()->debug($cot_enabled.json_encode($meta_keys), ['source' => 'save-meta-data']);
```

**WordPress post tables**

The is the result of running the snippet three times while using the WordPress post tables. All meta data entries added during the test appear both in `postmeta` and `wc_orders_meta`.

```
2023-08-03T16:05:34+00:00 DEBUG ["is_vat_exempt","_test_meta_data_1691078734_one","_test_meta_data_1691078734_two"]
2023-08-03T16:05:34+00:00 DEBUG [COT=off]_test_meta_data_1691078734_one is set
2023-08-03T16:05:34+00:00 DEBUG [COT=off]_test_meta_data_1691078734_two is set
2023-08-03T16:05:48+00:00 DEBUG ["is_vat_exempt","_test_meta_data_1691078734_one","_test_meta_data_1691078734_two","_test_meta_data_1691078748_one","_test_meta_data_1691078748_two"]
2023-08-03T16:05:48+00:00 DEBUG [COT=off]_test_meta_data_1691078748_one is set
2023-08-03T16:05:48+00:00 DEBUG [COT=off]_test_meta_data_1691078748_two is set
2023-08-03T16:05:58+00:00 DEBUG ["is_vat_exempt","_test_meta_data_1691078734_one","_test_meta_data_1691078734_two","_test_meta_data_1691078748_one","_test_meta_data_1691078748_two","_test_meta_data_1691078757_one","_test_meta_data_1691078757_two"]
2023-08-03T16:05:58+00:00 DEBUG [COT=off]_test_meta_data_1691078757_one is set
2023-08-03T16:05:58+00:00 DEBUG [COT=off]_test_meta_data_1691078757_two is set
```

**WooCommerce COT**

and this is the result of running the snippet three times while using COT. In this case, the meta data entries from the first two attempts appear both in `postmeta` and `wc_orders_meta`. All meta data entries appear in `wc_orders_meta`.

```
2023-08-03T16:12:16+00:00 DEBUG ["_billing_address_index","_shipping_address_index","is_vat_exempt"]
2023-08-03T16:12:16+00:00 DEBUG [COT=on]_test_meta_data_1691079136_one is NOT set
2023-08-03T16:12:16+00:00 DEBUG [COT=on]_test_meta_data_1691079136_two is NOT set
2023-08-03T16:12:28+00:00 DEBUG ["_billing_address_index","_shipping_address_index","is_vat_exempt","_test_meta_data_1691079136_one","_test_meta_data_1691079136_two"]
2023-08-03T16:12:28+00:00 DEBUG [COT=on]_test_meta_data_1691079148_one is NOT set
2023-08-03T16:12:28+00:00 DEBUG [COT=on]_test_meta_data_1691079148_two is NOT set
2023-08-03T16:12:34+00:00 DEBUG ["_billing_address_index","_shipping_address_index","is_vat_exempt","_test_meta_data_1691079136_one","_test_meta_data_1691079136_two","_test_meta_data_1691079148_one","_test_meta_data_1691079148_two"]
2023-08-03T16:12:34+00:00 DEBUG [COT=on]_test_meta_data_1691079153_one is NOT set
2023-08-03T16:12:34+00:00 DEBUG [COT=on]_test_meta_data_1691079153_two is NOT set
```

It seems that every snippet run can see the meta data entries added in previous runs, but not the new metadata entries added during the same run.

---

I'm not super familiar with running WooCommerce from this repo, so maybe I did something wrong and that's affecting my test results.

I checked out `fix/reduce_read_on_sync` and ran `pnpm run --filter='woocommerce' build` to prepare a copy of WooCommerce with the fix. I then copied the content of `plugins/woocommerce` to my local site and tried running the snippet again with the results I described above.
